### PR TITLE
Issue298

### PR DESCRIPTION
**Pull request by jlaber**
For this change, I didn't just allow overriding the default, I allow for shared configurations.  You could have many shared configurations, or just one default.  You really can do it however you want.  I also removed the creation of Health Indicators from the @PostConstruct and allow for registering post circuit breaker creation handling functions.  This was done so that if you create circuit breakers beyond the config file, you will still automatically get health indicators created for them.

Also:
This contains the changes from pull request: https://github.com/resilience4j/resilience4j/pull/312 - I needed the ability to override the defaults to build this functionality.  I fixed the issues that I mentioned in that PR.  One key note is that I did bump the spring boot 2 version to 2.1.x to allow for overriding the event consumers.  If we don't want that upgrade, I can change it back, and we just won't allow that overriding until we want to upgrade to 2.1.x

### CODE REVIEW

**Comment by Romeh** [comment]
@jlaber  can you please resolve the merge conflicts so we can play with ur PR ;)

**Comment by jlaber** [comment]
@Romeh Thanks for the notification.  I just merged in the conflicts. Please let me know what you think!

**Comment by RobWin** [comment]
Sry, could you rebase again ;)

**Comment by jlaber** [comment]
@RobWin - All merged in. Thanks.

File: resilience4j-spring/src/main/java/io/github/resilience4j/circuitbreaker/configure/CircuitBreakerConfiguration.java [hunk 1]
```diff
@@ -31,23 +32,23 @@
 @Configuration
 public class CircuitBreakerConfiguration {
 
+	private final CircuitBreakerConfigurationProperties circuitBreakerProperties;
+	private EventConsumerRegister eventConsumerRegister;
```
Comment (Romeh) [inline resilience4j-spring/src/main/java/io/github/resilience4j/circuitbreaker/configure/CircuitBreakerConfiguration.java]: I think EventConsumerRegister not needed to be class instance variable here

File: resilience4j-spring/src/main/java/io/github/resilience4j/circuitbreaker/configure/CircuitBreakerConfiguration.java [hunk 2]
```diff
@@ -61,4 +62,55 @@ public CircuitBreakerAspect circuitBreakerAspect(CircuitBreakerConfigurationProp
...
+     * 
+     * @param circuitBreakerRegistry The circuit breaker registry.
+     * @param eventConsumerRegistry The event consumer registry.
+     */
+    public void registerPostCreationEventConsumer(CircuitBreakerRegistry circuitBreakerRegistry,
+    		EventConsumerRegistry<CircuitBreakerEvent> eventConsumerRegistry) {
+    	eventConsumerRegister = new EventConsumerRegister(eventConsumerRegistry);
+		circuitBreakerRegistry.registerPostCreationConsumer(
+				(circuitBreaker, config) -> eventConsumerRegister.registerEventConsumer(circuitBreaker, config));
+    }
```
Comment (Romeh) [inline resilience4j-spring/src/main/java/io/github/resilience4j/circuitbreaker/configure/CircuitBreakerConfiguration.java]: you can use method reference eventConsumerRegister::registerEventConsumer for better readability

File: resilience4j-spring/src/main/java/io/github/resilience4j/circuitbreaker/configure/CircuitBreakerConfigurationProperties.java [hunk 3]
```diff
@@ -45,12 +48,28 @@ public void setCircuitBreakerAspectOrder(int circuitBreakerAspectOrder) {
     }
 
     private BackendProperties getBackendProperties(String backend) {
-        return backends.get(backend);
+    	BackendProperties properties = backends.get(backend);
+    	if(properties.getSharedConfigName() != null && !properties.getSharedConfigName().isEmpty() ) {
```
Comment (Romeh) [inline resilience4j-spring/src/main/java/io/github/resilience4j/circuitbreaker/configure/CircuitBreakerConfigurationProperties.java]: can we replace it with  if(StringUtils.isEmpty(backendProperties.getSharedConfigName())) { for better code readability

File: resilience4j-spring/src/main/java/io/github/resilience4j/circuitbreaker/configure/CircuitBreakerConfigurationProperties.java [hunk 4]
```diff
@@ -45,12 +48,28 @@ public void setCircuitBreakerAspectOrder(int circuitBreakerAspectOrder) {
...
     }
 
     public CircuitBreakerConfig createCircuitBreakerConfig(String backend) {
         return createCircuitBreakerConfig(getBackendProperties(backend));
     }
+    
+    public CircuitBreakerConfig createCircuitBreakerConfigFromShared(String sharedConfig) {
+        BackendProperties backendProperties = getSharedConfigProperties(sharedConfig);
+        if(backendProperties.getSharedConfigName() == null) {
+            return buildCircuitBreakerConfig(backendProperties).configurationName(sharedConfig).build();
```
Comment (Romeh) [inline resilience4j-spring/src/main/java/io/github/resilience4j/circuitbreaker/configure/CircuitBreakerConfigurationProperties.java]: same remark as above

File: resilience4j-spring-boot/src/main/java/io/github/resilience4j/circuitbreaker/autoconfigure/CircuitBreakerConfigurationOnMissingBean.java [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    @ConditionalOnMissingBean
+    public CircuitBreakerRegistry circuitBreakerRegistry(EventConsumerRegistry<CircuitBreakerEvent> eventConsumerRegistry) {
+        CircuitBreakerRegistry circuitBreakerRegistry = CircuitBreakerRegistry.ofDefaults();
+
+        // Register the event consumers
+        circuitBreakerConfiguration.registerPostCreationEventConsumer(circuitBreakerRegistry, eventConsumerRegistry);
+        // Register a consumer to hook up any health indicators for circuit breakers after creation. This will catch ones that get
+        // created beyond initially configured backends.
+        circuitBreakerRegistry.registerPostCreationConsumer((circuitBreaker, config) -> createHeathIndicatorForCircuitBreaker(circuitBreaker, config));
+
```
Comment (Romeh) [inline resilience4j-spring-boot/src/main/java/io/github/resilience4j/circuitbreaker/autoconfigure/CircuitBreakerConfigurationOnMissingBean.java]: can we use method reference here this::createHeathIndicatorForCircuitBreaker

File: resilience4j-spring-boot2/src/main/java/io/github/resilience4j/circuitbreaker/autoconfigure/CircuitBreakerConfigurationOnMissingBean.java [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+    @Bean
+    @ConditionalOnMissingBean
+    public CircuitBreakerRegistry circuitBreakerRegistry(EventConsumerRegistry<CircuitBreakerEvent> eventConsumerRegistry) {
+        CircuitBreakerRegistry circuitBreakerRegistry = CircuitBreakerRegistry.ofDefaults();
+
+        // Register the event consumers
+        circuitBreakerConfiguration.registerPostCreationEventConsumer(circuitBreakerRegistry, eventConsumerRegistry);
+        // Register a consumer to hook up any health indicators for circuit breakers after creation. This will catch ones that get
+        // created beyond initially configured backends.
+        circuitBreakerRegistry.registerPostCreationConsumer((circuitBreaker, config) -> createHeathIndicatorForCircuitBreaker(circuitBreaker, config));
```
Comment (Romeh) [inline resilience4j-spring-boot2/src/main/java/io/github/resilience4j/circuitbreaker/autoconfigure/CircuitBreakerConfigurationOnMissingBean.java]: can we use method reference here this::createHeathIndicatorForCircuitBreaker

File: resilience4j-spring/src/main/java/io/github/resilience4j/circuitbreaker/configure/CircuitBreakerConfiguration.java [hunk 7]
```diff
@@ -61,4 +62,55 @@ public CircuitBreakerAspect circuitBreakerAspect(CircuitBreakerConfigurationProp
...
     }
+    
+    /**
+     * Initializes the backends configured in the properties.
+     * 
+     * @param circuitBreakerRegistry The circuit breaker registry.
+     * @param eventConsumerRegistry The event consumer registry.
+     */
+    public void initializeBackends(CircuitBreakerRegistry circuitBreakerRegistry,
+    							   EventConsumerRegistry<CircuitBreakerEvent> eventConsumerRegistry) {
```
Comment (Romeh) [inline resilience4j-spring/src/main/java/io/github/resilience4j/circuitbreaker/configure/CircuitBreakerConfiguration.java]: EventConsumerRegistry<CircuitBreakerEvent> eventConsumerRegistry is unneeded dependency here

**Comment by Romeh** [comment]
@jlaber  i had small code comments there , can you please update them plus can u please add in spring boot 1 doc that we are not overriding Event consumer and finally add a some documentation to rate limiter that spring beans can be overridden there as well :) .
 
@RobWin are we going to base our-self on spring boot 2.1.* for spring boot 2 starter ?

**Comment by RobWin** [comment]
@Romeh I already updated master to use Spring Boot 2.1.3.RELEASE

**Comment by jlaber** [comment]
@Romeh or @RobWin Are either of you familiar with that Spring Boot Asnyc Rety test that i have failing:
io.github.resilience4j.retry.RetryAutoConfigurationTest > testRetryAutoConfigurationAsync FAILED

I can't reproduce that at all when I run the tests.  Does it have something to do with the Spring Context not cleaning up for it appropriately? I see the Spring Boot 2 version has a @DirtiesContext which isn't available for the Spring Boot 1.

**Comment by RobWin** [comment]
Yes I can fix it tomorrow

**Comment by Romeh** [comment]
@jlaber the test is fixed now on master , please re-base ur PR and also resolve the merge conflicts then i guess we will be good to GO :)

**Comment by jlaber** [comment]
@Romeh All merged in.

**Comment by jlaber** [comment]
@Romeh @RobWin 
The build is failing again on one of those Retry tests:

io.github.resilience4j.retry.RetryAutoConfigurationTest > testRetryAutoConfigurationAsync FAILED
    org.junit.ComparisonFailure at RetryAutoConfigurationTest.java:146
2019-04-04 13:26:23.702  INFO 6356 --- [     Thread-108] o.s.s.concurrent.ThreadPoolTaskExecutor  : Shutting down ExecutorService 'applicationTaskExecutor'

I seem to only see these out on the build server and seem to be sporadic. Locally these errors don't occur.

Also, I was just wondering if you think this is a good PR in general and functionality is good? If so, would you please consider pulling the PR in.  Because of the number of changes in the configuration files, it can be tough to keep up-to-date with the merge conflicts that come in from all the other new functionality. I don't want to accidentally break anything new because of missing something trying to resolve the merge.

**Comment by RobWin** [comment]
Yes, we are working again on fixing this test.
We will first release 0.14.0 and then take us time to merge this PR.

**Comment by jlaber** [comment]
Ok, cool.  Thanks for the info.

**Comment by jlaber** [comment]
Hi @Romeh @RobWin,

It's great to see that this got merged in! I was just wondering what might have happened because it doesn't look like I show up as a contributor to the project?

Thanks,
Jason

**Comment by RobWin** [comment]
Good question. Btw the Apache License in your added files seems to be copied or?
You name is not Slawomir Kowalski, correct?

**Comment by RobWin** [comment]
Good questions, seems mange people contributed commits and I always squash a PR before it is merged. Maybe Github uses the first commit to choose the contributor?

**Comment by RobWin** [comment]
But maybe you could update our Spring Boot documentation and I merge it so that you are also listed.

**Pull request merged**